src/rust/engine/fs/src/glob_matching.rs

```diff
@@ -0,0 +1,397 @@
+          .map(|entry| entry.input)
+          .filter(|parsed_source| !inputs_with_matches.contains(parsed_source))
+          .collect();
+
+        if !non_matching_inputs.is_empty() {
+          // TODO(#5684): explain what global and/or target-specific option to set to
+          // modify this behavior!
+          let msg = format!(
+            "Globs did not match. Excludes were: {:?}. Unmatched globs were: {:?}.",
+            exclude.exclude_patterns(),
+            non_matching_inputs
+              .iter()
+              .map(|parsed_source| parsed_source.0.clone())
+              .collect::<Vec<_>>(),
+          );
+          if strict_match_behavior.should_throw_on_error() {
+            return future::err(Self::mk_error(&msg));
+          } else {
+            // TODO(#5683): this doesn't have any useful context (the stack trace) without
+            // being thrown -- this needs to be provided, otherwise this is unusable.
+            warn!("{}", msg);
+          }
+        }
+      }
+
+      future::ok(match_results)
+    })
+      .to_boxed()
+  }
+
+  ///
+  /// Apply a PathGlob, returning PathStats and additional PathGlobs that are needed for the
+  /// expansion.
+  ///
+  fn expand_single(
+    &self,
+    sourced_glob: GlobWithSource,
+    exclude: &Arc<GitignoreStyleExcludes>,
+  ) -> BoxFuture<SingleExpansionResult, E> {
+    match sourced_glob.path_glob.clone() {
+      PathGlob::Wildcard { canonical_dir, symbolic_path, wildcard } =>
+      // Filter directory listing to return PathStats, with no continuation.
+        self.directory_listing(canonical_dir, symbolic_path, wildcard, exclude)
+        .map(move |path_stats| SingleExpansionResult {
+          sourced_glob,
+          path_stats,
+          globs: vec![],
+        })
+        .to_boxed(),
+      PathGlob::DirWildcard { canonical_dir, symbolic_path, wildcard, remainder } =>
+      // Filter directory listing and request additional PathGlobs for matched Dirs.
+        self.directory_listing(canonical_dir, symbolic_path, wildcard, exclude)
+        .and_then(move |path_stats| {
+          path_stats.into_iter()
+            .filter_map(|ps| match ps {
+              PathStat::Dir { path, stat } =>
+                Some(
+                  PathGlob::parse_globs(stat, path, &remainder)
+                    .map_err(|e| Self::mk_error(e.as_str()))
+                ),
+              PathStat::File { .. } => None,
+            })
+            .collect::<Result<Vec<_>, E>>()
+        })
+        .map(move |path_globs| {
+          let flattened = path_globs
+            .into_iter()
+            .flat_map(|path_globs| path_globs.into_iter())
+            .collect();
+          SingleExpansionResult {
+            sourced_glob,
+            path_stats: vec![],
+            globs: flattened,
+          }
+        })
+        .to_boxed(),
+    }
+  }
+
+  ///
+  /// Canonicalize the Link for the given Path to an underlying File or Dir. May result
```

Comment:
`canonicalize` feels like a `VFS` thing not a `GlobMatchingImplementation` thing? I didn't realise we implemented it by expanding globs, that seems a little weird...

Reply:
It needs to use a symlink-aware codepath in order to properly track symlink chains.

Reply:
My assumption was that we'd repeatedly call `read_link`, rather than repeatedly making `PathGlobs` and calling `expand` :)

Reply:
More than that: you need to do it for every path component up from the root. So that `/blah/this_is_a_link/foo/this_is_also_a_link/bar` is tracked correctly.

Reply:
> More than that: you need to do it for every path component up from the root. So that `blah/this_is_a_link/foo/this_is_also_a_link/bar` is tracked correctly.

This feels like something `read_link` could do itself?

Reply:
Maybe. I expect that it would lead to redundant expansion of pathglobs. But let's save refactoring for a separate PR.

Reply:
Oh! I didn't realize we `canonicalize()`d in `directory_listing()`. With that I think it's fine.